File: Triangle.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Fri Oct 18 10:19:14 2019

@author: tanvirkaur
"""

# time complexity = O(n), n is the number of elements the triangle
# space complexity = O(n)
# we are moving  from the second last row to the 0th row and returning the value of top of the triangle
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        A = len(triangle)
        for i in range(A-2,-1,-1):
            for j in range(0,i+1,+1):
                triangle[i][j] = triangle[i][j] + min(triangle[i+1][j], triangle[i+1][j+1])
        return triangle[0][0]
             

File: Triangle.java
package DynamicProgramming8;

import java.util.List;

public class Triangle {
    /* Created by palak on 8/16/2021 */

    /**
     Brute Force:
     Maintain a list at every level.
     Start from the last level. We can store the minimim elements at each level.
     But then we can also store the indices instead of elements.

     Bottom Up approch: Take the minimum from the pair and add it to the previous level value
     Until you reach the first element.

     Time complexity: O(n)
     Space Complexity: O(n)
     */
    public int minimumTotal(List<List<Integer>> triangle) {
        if(triangle == null || triangle.get(0).size() == 0 || triangle.size() == 0) return 0;
        int n = triangle.size();
        int dp[] = new int[n];
        for(int i = 0 ; i < dp.length ; i++) {
            dp[i] = triangle.get(n - 1).get(i);
        }
        for(int i = n - 2 ; i >= 0 ; i--) {
            //i = triangle.get(i).size()
            for(int j = 0 ; j <= i ; j++) {
                dp[j] = triangle.get(i).get(j) + Math.min(dp[j], dp[j+1]);
            }
        }
        return dp[0];
    }

    public static void main(String[] args) {

    }
}
